gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.BackendsApiManager
+import io.glutenproject.extension.columnar.TransformHints
+import io.glutenproject.utils.PhysicalPlanSelector
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Final, Partial}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.aggregate.{BaseAggregateExec, HashAggregateExec, ObjectHashAggregateExec, SortAggregateExec}
+
+/**
+ * Merge two phase hash-based aggregate into one aggregate in the spark plan if there is no shuffle:
+ *
+ * Merge HashAggregate(t1.i, SUM, final) + HashAggregate(t1.i, SUM, partial) into
+ * HashAggregate(t1.i, SUM, complete)
+ *
+ * Note: this rule must be applied before the `PullOutPreProject` rule, because the
+ * `PullOutPreProject` rule will modify the attributes in some cases.
+ */
+case class MergeTwoPhasesHashBaseAggregate(session: SparkSession) extends Rule[SparkPlan] {
+
+  val columnarConf: GlutenConfig = GlutenConfig.getConf
+  val scanOnly: Boolean = columnarConf.enableScanOnly
+  val enableColumnarHashAgg: Boolean = !scanOnly && columnarConf.enableColumnarHashAgg
+  val replaceSortAggWithHashAgg = BackendsApiManager.getSettings.replaceSortAggWithHashAgg
+
+  private def isTransformable(agg: BaseAggregateExec): Boolean = {
+    if (!TransformHints.isAlreadyTagged(agg)) {
+      // Check whether the aggregate is transformable.
+      // Note: do not use the AddTransformHintRule to apply all plan again
+      val transformer = BackendsApiManager.getSparkPlanExecApiInstance
+        .genHashAggregateExecTransformer(
+          agg.requiredChildDistributionExpressions,
+          agg.groupingExpressions,
+          agg.aggregateExpressions,
+          agg.aggregateAttributes,
+          0, // there is no `initialInputBufferOffset` field in Spark 3.2
+          agg.resultExpressions,
+          agg.child
+        )
+      transformer.doValidate().isValid
```

Comment:
Here we have not pulled out pre/post project, so this validation seems very likely to fail. Do we need this check ? I think it should be fine even we fallback to vanilla Spark after merging aggregates, since there already exists a rule ReplaceHashWithSortAgg to do the similar thing.

Reply:
I am not sure that merging sort aggregates also works for the vanilla Spark, do you have any idea for this? the hash based aggregates may be work even fallback.

Reply:
removed